File: cli/internal/backend_hosted.go
package internal

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

var ErrHostedTokenInvalid = errors.New("hosted token is invalid or expired; run daily-tasks login again")
var ErrHostedTokenMissing = errors.New("hosted token missing; run daily-tasks login")

type HostedBackend struct {
	APIURL string
	Token  string
	Client *http.Client
}

type hostedSyncPayload struct {
	Data       string `json:"data"`
	History    string `json:"history"`
	remoteETag string
}

func NewHostedBackend(apiURL, token string) *HostedBackend {
	return &HostedBackend{
		APIURL: strings.TrimRight(strings.TrimSpace(apiURL), "/"),
		Token:  strings.TrimSpace(token),
	}
}

func (b *HostedBackend) httpClient() *http.Client {
	if b.Client != nil {
		return b.Client
	}
	return &http.Client{Timeout: 15 * time.Second}
}

func (b *HostedBackend) syncURL() string {
	return strings.TrimRight(b.APIURL, "/") + "/api/v1/sync"
}

func (b *HostedBackend) Fetch(key string) (Blob, error) {
	payload, err := b.fetchPayload()
	if err != nil {
		return Blob{}, err
	}
	body, err := payload.bytesFor(key)
	if err != nil {
		return Blob{}, err
	}
	return Blob{Bytes: body, Etag: payload.etag()}, nil
}

func (b *HostedBackend) Push(key string, body []byte, ifMatch string) error {
	payload, err := b.fetchPayload()
	if errors.Is(err, ErrRemoteNotFound) {
		payload = hostedSyncPayload{}
	} else if err != nil {
		return err
	}
	if err := payload.setBytes(key, body); err != nil {
		return err
	}
	return b.putPayload(payload, ifMatch)
}

func (b *HostedBackend) fetchPayload() (hostedSyncPayload, error) {
	req, err := http.NewRequest(http.MethodGet, b.syncURL(), nil)
	if err != nil {
		return hostedSyncPayload{}, err
	}
	b.authorize(req)
	resp, err := b.httpClient().Do(req)
	if err != nil {
		return hostedSyncPayload{}, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
		return hostedSyncPayload{}, ErrHostedTokenInvalid
	}
	if resp.StatusCode == http.StatusNotFound {
		return hostedSyncPayload{}, ErrRemoteNotFound
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return hostedSyncPayload{}, fmt.Errorf("hosted backend status %d", resp.StatusCode)
	}
	var payload hostedSyncPayload
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return hostedSyncPayload{}, err
	}
	payload.remoteETag = resp.Header.Get("ETag")
	return payload, nil
}

func (b *HostedBackend) putPayload(payload hostedSyncPayload, ifMatch string) error {
	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	req, err := http.NewRequest(http.MethodPut, b.syncURL(), bytes.NewReader(body))
	if err != nil {
		return err
	}
	b.authorize(req)
	req.Header.Set("Content-Type", "application/json")
	setIfMatchHeader(req, ifMatch)
	resp, err := b.httpClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
		_, _ = io.Copy(io.Discard, resp.Body)
		return ErrHostedTokenInvalid
	}
	if resp.StatusCode == http.StatusPreconditionFailed {
		return ErrEtagMismatch
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("hosted push failed with status %d", resp.StatusCode)
	}
	return nil
}

func (b *HostedBackend) authorize(req *http.Request) {
	if b.Token != "" {
		req.Header.Set("Authorization", "Bearer "+b.Token)
	}
}

func (p hostedSyncPayload) bytesFor(key string) ([]byte, error) {
	switch key {
	case KeyData:
		return decodeHostedBlob(p.Data, key)
	case KeyHistory:
		return decodeHostedBlob(p.History, key)
	default:
		return nil, fmt.Errorf("hosted backend: unknown key %q", key)
	}
}

func (p *hostedSyncPayload) setBytes(key string, body []byte) error {
	encoded := base64.StdEncoding.EncodeToString(body)
	switch key {
	case KeyData:
		p.Data = encoded
	case KeyHistory:
		p.History = encoded
	default:
		return fmt.Errorf("hosted backend: unknown key %q", key)
	}
	return nil
}

func (p hostedSyncPayload) etag() string {
	if p.remoteETag != "" {
		return p.remoteETag
	}
	sum := sha256.Sum256([]byte(p.Data + "\x00" + p.History))
	return `"hosted-` + hex.EncodeToString(sum[:]) + `"`
}

func decodeHostedBlob(encoded, key string) ([]byte, error) {
	if encoded == "" {
		return nil, ErrRemoteNotFound
	}
	body, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil, fmt.Errorf("hosted backend: invalid %s payload: %w", key, err)
	}
	return body, nil
}

File: mobile/src/services/backend_hosted.ts
import { encode as base64Encode, decode as base64Decode } from 'base-64';
import * as AuthSession from 'expo-auth-session';
import * as WebBrowser from 'expo-web-browser';
import { hostedApiUrl } from '../config/env';
import {
  Backend,
  Blob,
  EtagMismatchError,
  IF_NONE_MATCH_ANY,
  KeyData,
  KeyHistory,
} from './backend';

WebBrowser.maybeCompleteAuthSession();

export type HostedProvider = 'google' | 'facebook';

export type HostedSettings = {
  apiUrl: string;
  token?: string;
  email?: string;
};

export type HostedLoginResult = {
  token: string;
  email?: string;
};

type HostedSyncPayload = {
  data?: string;
  history?: string;
  updated_at?: string;
};

export class HostedAuthError extends Error {
  constructor(message = 'Hosted login expired. Sign in again.') {
    super(message);
    this.name = 'HostedAuthError';
  }
}

export const defaultHostedSettings: HostedSettings = {
  apiUrl: hostedApiUrl,
};

function normalizeApiUrl(apiUrl: string): string {
  return apiUrl.trim().replace(/\/+$/, '');
}

function syncUrl(apiUrl: string): string {
  return `${normalizeApiUrl(apiUrl)}/api/v1/sync`;
}

function authHeader(token: string): Record<string, string> {
  return { Authorization: `Bearer ${token}` };
}

function decodeBlob(encoded: string | undefined, key: string): string | null {
  if (!encoded) {
    return null;
  }
  try {
    return base64Decode(encoded);
  } catch (err) {
    throw new Error(`hosted backend: invalid ${key} payload: ${(err as Error).message}`);
  }
}

function encodeBlob(body: string): string {
  return base64Encode(body);
}

function etagFromResponse(res: Response): string {
  return res.headers.get('ETag') ?? '';
}

function applyIfMatch(headers: Record<string, string>, ifMatch?: string): void {
  if (!ifMatch) {
    return;
  }
  if (ifMatch === IF_NONE_MATCH_ANY) {
    headers['If-None-Match'] = '*';
  } else {
    headers['If-Match'] = ifMatch;
  }
}

export class HostedBackend implements Backend {
  readonly settings: HostedSettings;

  constructor(settings: HostedSettings) {
    this.settings = {
      ...settings,
      apiUrl: normalizeApiUrl(settings.apiUrl || hostedApiUrl),
      token: settings.token?.trim(),
    };
  }

  private requireToken(): string {
    if (!this.settings.token) {
      throw new HostedAuthError('Hosted token missing. Sign in again.');
    }
    return this.settings.token;
  }

  private async fetchPayload(): Promise<{ payload: HostedSyncPayload; etag: string }> {
    const res = await globalThis.fetch(syncUrl(this.settings.apiUrl), {
      headers: {
        ...authHeader(this.requireToken()),
        Accept: 'application/json',
      },
    });
    if (res.status === 401 || res.status === 403) {
      throw new HostedAuthError();
    }
    if (!res.ok) {
      throw new Error(`Hosted fetch failed: ${res.status}`);
    }
    return { payload: await res.json() as HostedSyncPayload, etag: etagFromResponse(res) };
  }

  async fetch(key: string): Promise<Blob> {
    const { payload, etag } = await this.fetchPayload();
    switch (key) {
      case KeyData:
        return { bytes: decodeBlob(payload.data, key), etag };
      case KeyHistory:
        return { bytes: decodeBlob(payload.history, key), etag };
      default:
        throw new Error(`hosted backend: unknown key ${key}`);
    }
  }

  async push(key: string, body: string, ifMatch?: string): Promise<void> {
    const { payload } = await this.fetchPayload();
    switch (key) {
      case KeyData:
        payload.data = encodeBlob(body);
        break;
      case KeyHistory:
        payload.history = encodeBlob(body);
        break;
      default:
        throw new Error(`hosted backend: unknown key ${key}`);
    }

    const headers: Record<string, string> = {
      ...authHeader(this.requireToken()),
      'Content-Type': 'application/json',
    };
    applyIfMatch(headers, ifMatch);
    const res = await globalThis.fetch(syncUrl(this.settings.apiUrl), {
      method: 'PUT',
      headers,
      body: JSON.stringify({
        data: payload.data ?? encodeBlob('{}'),
        history: payload.history ?? encodeBlob('{}'),
      }),
    });
    if (res.status === 401 || res.status === 403) {
      throw new HostedAuthError();
    }
    if (res.status === 412) {
      throw new EtagMismatchError();
    }
    if (!res.ok) {
      throw new Error(`Hosted save failed: ${res.status}`);
    }
  }
}

export function buildHostedAuthUrl(provider: HostedProvider, apiUrl: string, redirectUri: string): string {
  const params = new URLSearchParams();
  params.set('redirect_uri', redirectUri);
  return `${normalizeApiUrl(apiUrl)}/auth/${provider}?${params.toString()}`;
}

export function parseHostedLoginCallback(url: string): HostedLoginResult {
  const parsed = new URL(url);
  const token = parsed.searchParams.get('token');
  if (!token) {
    throw new Error('Hosted login did not return a token');
  }
  return {
    token,
    email: parsed.searchParams.get('email') || undefined,
  };
}

export async function startHostedLogin(provider: HostedProvider, apiUrl = hostedApiUrl): Promise<HostedLoginResult> {
  const redirectUri = AuthSession.makeRedirectUri({ scheme: 'daily-tasks', path: 'auth' });
  const authUrl = buildHostedAuthUrl(provider, apiUrl, redirectUri);
  const result = await WebBrowser.openAuthSessionAsync(authUrl, redirectUri);
  if (result.type !== 'success' || !result.url) {
    throw new Error('Hosted login was cancelled');
  }
  return parseHostedLoginCallback(result.url);
}
